### PR TITLE
TDL-13631: Added include_deleted configuration

### PR DESCRIPTION
**Pull request by dbshah1212**
# Description of change
Added include delete configuration

# Manual QA steps
 - Ran sync mode without include_delete and getting more records and ran sync mode with include_delete=False and getting fewer records for invoices 
 
# Risks
 - 
 
# Rollback steps
 - revert this branch

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -51,6 +52,8 @@ This tap:
 
    The `site` parameter represents the name of your specific Chargebee site (e.g. `https://{site}.chargebee.com/api/v2/subscriptions`)
 
+   The `include_deleted` is optional flag to ask do you want deleted records of all stream or not. Default it's true.
```
Comment (karanpanchal-crest) [inline README.md:55]: `The `include_deleted` is an optional flag to ask if you want deleted records of all streams or not. Default: true`

File: tap_chargebee/streams/base.py [hunk 2]
```diff
@@ -175,23 +175,24 @@ def sync_data(self):
             
             to_write = self.get_stream_data(records)
             
-            if self.ENTITY == 'event':
+            if self.config.get('include_deleted') not in ['false','False', False] and self.ENTITY == 'event':
```
Comment (karanpanchal-crest) [inline tap_chargebee/streams/base.py:178]: Keep self.entity == event in  the nested check

File: tap_chargebee/streams/base.py [hunk 3]
```diff
@@ -175,23 +175,24 @@ def sync_data(self):
...
-            if self.ENTITY == 'plan':
-                for plan in Util.plans:
-                    to_write.append(plan)
-            if self.ENTITY == 'addon':
-                for addon in Util.addons:
-                    to_write.append(addon)
-            if self.ENTITY == 'coupon':
-                for coupon in Util.coupons:
-                    to_write.append(coupon) 
+            if self.config.get('include_deleted') not in ['false','False', False]:
```
Comment (karanpanchal-crest) [inline tap_chargebee/streams/base.py:186]: Take this code in the nested scope of the above condition

File: tests/test_chargebee_include_delete.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    def get_properties(self):
+        properties = super().get_properties()
+
+        # include_deleted is an optional property for configuration
+        if self.include_deleted is False:
+            properties["include_deleted"] = self.include_deleted
+
+        return properties
+
+    def run_sync(self, expected_streams):
```
Comment (karanpanchal-crest) [inline tests/test_chargebee_include_delete.py:29]: These test cases are obsolete. Please update it as per the latest test cases added in the tap

File: tests/test_chargebee_include_delete.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+        synced_records_with_include_deleted_false = self.run_sync(
+            expected_streams)
+
+        # For include_delete false
+        self.include_deleted = False
+
+        synced_records_with_include_deleted_true = self.run_sync(
+            expected_streams)
+
+        self.assertGreater(sum(synced_records_with_include_deleted_false.values()), sum(
```
Comment (karanpanchal-crest) [inline tests/test_chargebee_include_delete.py:60]: Assert explicitly if the events contains 'deleted' records

Comment (karanpanchal-crest) [reply tests/test_chargebee_include_delete.py:60]: Reference: https://github.com/singer-io/tap-xero/blob/master/tests/test_archived_contacts_support.py

File: README.md (hunk 1 above)
Comment (dbshah1212) [reply README.md:55]: updated

File: tap_chargebee/streams/base.py (hunk 2 above)
Comment (dbshah1212) [reply tap_chargebee/streams/base.py]: DOne

File: tap_chargebee/streams/base.py (hunk 3 above)
Comment (dbshah1212) [reply tap_chargebee/streams/base.py:178]: Done

File: tests/test_chargebee_include_delete.py (hunk 5 above)
Comment (dbshah1212) [reply tests/test_chargebee_include_delete.py]: Won't possible here, as we are not getting deleted records for include_deleted=false. So the only difference will work here.

**Review by karanpanchal-crest — APPROVED** [review]

File: tap_chargebee/client.py [hunk 6]
```diff
@@ -29,6 +29,9 @@ def __init__(self, config, api_result_limit=100, include_deleted=True):
         self.include_deleted = include_deleted
         self.user_agent = self.config.get('user_agent')
 
+        if self.config.get('include_deleted') in ['false','False', False]:
+            self.include_deleted = False
```
Comment (KAllan357) [inline tap_chargebee/client.py:33]: This seems odd. Why is this saved onto the client and never referenced?

File: tap_chargebee/streams/base.py [hunk 7]
```diff
@@ -175,23 +175,24 @@ def sync_data(self):
...
```
Comment (KAllan357) [inline tap_chargebee/streams/base.py:178]: I'm not sure why this behavior is behind the new flag. Shouldn't the flag only affect this logic by appending the events when they're "*_deleted" and the flag is true?

File: tap_chargebee/client.py (hunk 6 above)
Comment (dbshah1212) [reply tap_chargebee/client.py:33]: We haven't added this parameter, it was there with the default value `true`. We just updated it for the particular scenario.  
And it got referred at line no 49 inside `get_params` function:  `params['include_deleted'] = self.include_deleted`

File: tap_chargebee/streams/base.py (hunk 7 above)
Comment (dbshah1212) [reply tap_chargebee/streams/base.py:206]: Yes, it should like that. But for some of the streams `include_deleted` param isn't supported by API, so for those streams, tap collecting deleted records from the Event stream. 

If include_deleted=False then we won't need to collect these records from the event stream, hence I put this behavior behind the flag.

**Pull request merged**